### ISSUE TITLE
How to use a different base uri for operations

### ISSUE THREAD

**Issue by jmerkow** [Azure/autorest#4245]
I have an API that receives tokens from `https://management.azure.com`, but the operations are on a different endpoint that something like `$"https://{resourceRegion}.xyz.net/"`.  I have been handling this by adding a property to the partial classthen manually changing the generated code

```csharp
public partial class MyManagementClient
{
    public Uri OperationEndpoint { get; set; }
}
```

```diff
public MyManagementClient(Uri endpoint, TokenCredential tokenCredential, LabelingManagementClientOptions options = null)
{
    endpoint ??= new Uri("https://management.azure.com");

    options ??= new MyManagementClientOptions();
    _clientDiagnostics = new ClientDiagnostics(options);
    _pipeline = ManagementPipelineBuilder.Build(tokenCredential, endpoint, options);
    _endpoint = endpoint;
}

/// <summary> Returns an instance of MyOperations. </summary>
- public virtual MainOperations Main => new MainOperations(_clientDiagnostics, _pipeline, _endpoint);
+ public virtual MainOperations Main => new MainOperations(_clientDiagnostics, _pipeline, OperationEndpoint);
```

Obviously, this isn't ideal since I am editing the generated code....  I cannot replace the property, (`The type 'MyManagementClient' already contains a definition for 'Main'`

Is there a way to do directly with autorest?

**Comment by timotheeguerin** [comment]
Hey, not sure I fully get what you are saying?

Are you just trying to change the default host being  `https://management.azure.com`? If so does the openapi `servers` property not solve this?
https://swagger.io/docs/specification/api-host-and-base-path/

**Comment by jmerkow** [comment]
It might, I was not aware of it.  This functionality is OpenAPI 3.0 only correct (if I have the region in the url)? So I need to upgrade my swagger...

essentially the ManagementPipelineBuilder needs to be `new Uri("https://management.azure.com");` which is where I get auth tokens but the operations are on a different base URL (based on region of the resource).

**Comment by timotheeguerin** [comment]
what do you mean by you use management.azure.com to get the auth tokens? Would recommend not to defint that in the swagger then as this is an external endpoint. Autorest also provide a built in authentication if you use AzureAD or key based authentication

OpenAPI 3 adds the ability to have multiple servers but you can still choose the host in swagger 2.0 https://swagger.io/docs/specification/2-0/api-host-and-base-path/

**Comment by jmerkow** [comment]
I think I am going about this wrong.  Let me get more specific.  I am working with AzureML services to interact with resources (i.e Workspaces), then I need to switch to a rest API specific to that workspace location to interact with resources within that workspace, (see [docs](https://docs.microsoft.com/en-us/azure/machine-learning/how-to-manage-rest)).

In rest I would do 
```bash
curl https://management.azure.com/subscriptions/{your-subscription-id}/resourceGroups/{your-resource-group}/providers/Microsoft.MachineLearningServices/workspaces/?api-version=2021-03-01-preview \
-H "Authorization:Bearer {my-access-token}"
```
With this, I am able to obtain the base url for my other API, in my case that url is `"labeling": "https://{resourceRegion}.experiments.azureml.net"`. The second API needs to authenticate with `https://management.azure.com/.default` to get an bearer token, but the rest of the communication would be though at `"https://{resourceRegion}.experiments.azureml.net"`

So in my custom-constructor I am passing setting 
```csharp
_pipeline = ManagementPipelineBuilder.Build(tokenCredential, new Uri("https://management.azure.com"), options);
_endpoint = new Uri($"https://{resourceRegion}.experiments.azureml.net/");
```

Since  `HttpPipelineBuilder` builds like this:
```csharp
HttpPipelineBuilder.Build(options, new BearerTokenAuthenticationPolicy(credential, $"{endpoint}/.default"))` 
```
I get proper authentication for the other API endpoint.

**Comment by timotheeguerin** [comment]
I see, is there no provided Azure SDK for the 2 endpoint already(not familiar with this service)?

If not I see 2 ways:
1. You upgrade to openapi3 and use the servers override to change the endpoints.
2. You generate 2 seperate sdk for the different endpoints

**Comment by timotheeguerin** [comment]
1. Yeah that's what I meant, make sense.
2. Actually testing that it seems like the autorest.csharp generator doesn't support that yet. Tested with that:
```
openapi: 3.0.1
info:
  title: test
  version: v1.0
servers:
  - url: https://default.com
    description: Default server
paths:
  /default:
    get:
      operationId: getDefault
      responses:
        "200":
          description: OK
  /other:
    get:
      operationId: getOther
      responses:
        "200":
          description: OK
    servers:
      - url: https://other.com
        description: Default server
```
Feel free to file an issue on the csharp repo for this.

So I would recommend you just create 2 specs and generate 2 separate SDK for both endpoints. For the authentication you can just add --security to configure AAD auth (see https://github.com/Azure/autorest/blob/main/docs/generate/authentication.md)

**Comment by jmerkow** [comment]
I think all is to add security, does this conflict with `azure-arm: true`? When I remove `azure-arm: true` and add `security: true` to the readme I get a bunch of clients instead of one, this doesn't really seem to be what I want either.

**Comment by timotheeguerin** [comment]
Yeah not fully sure what is the behavior for azure-arm in the csharp generator, might be better to ask the question in the autorest.csharp repo about the sepecificity of what you want for the client.
